google-cloud-pubsublite/src/test/java/com/google/cloud/pubsublite/internal/BlockingPullSubscriberImplTest.java

```diff
@@ -0,0 +1,233 @@
+      SeekRequest.newBuilder()
+          .setCursor(Cursor.newBuilder().setOffset(initialOffset.value()))
+          .build();
+  private final FlowControlSettings flowControlSettings =
+      FlowControlSettings.builder().setBytesOutstanding(10).setMessagesOutstanding(20).build();
+  // Initialized in setUp.
+  private BlockingPullSubscriberImpl subscriber;
+  private Consumer<ImmutableList<SequencedMessage>> messageConsumer;
+  private ApiService.Listener errorListener;
+  private final ExecutorService executorService = Executors.newCachedThreadPool();
+
+  @Before
+  public void setUp() throws Exception {
+    when(underlying.startAsync()).thenReturn(underlying);
+    SeekRequest seek =
+        SeekRequest.newBuilder()
+            .setCursor(Cursor.newBuilder().setOffset(initialOffset.value()).build())
+            .build();
+    when(underlying.seek(seek)).thenReturn(ApiFutures.immediateFuture(initialOffset));
+    FlowControlRequest flow =
+        FlowControlRequest.newBuilder()
+            .setAllowedBytes(flowControlSettings.bytesOutstanding())
+            .setAllowedMessages(flowControlSettings.messagesOutstanding())
+            .build();
+    when(underlyingFactory.newSubscriber(any()))
+        .thenAnswer(
+            args -> {
+              messageConsumer = args.getArgument(0);
+              return underlying;
+            });
+    doAnswer(
+            (Answer<Void>)
+                args -> {
+                  errorListener = args.getArgument(0);
+                  return null;
+                })
+        .when(underlying)
+        .addListener(any(), any());
+
+    subscriber =
+        new BlockingPullSubscriberImpl(underlyingFactory, flowControlSettings, initialSeek);
+
+    InOrder inOrder = inOrder(underlyingFactory, underlying);
+    inOrder.verify(underlyingFactory).newSubscriber(any());
+    inOrder.verify(underlying).addListener(any(), any());
+    inOrder.verify(underlying).startAsync();
+    inOrder.verify(underlying).awaitRunning();
+    inOrder.verify(underlying).seek(seek);
+    inOrder.verify(underlying).allowFlow(flow);
+
+    assertThat(messageConsumer).isNotNull();
+    assertThat(errorListener).isNotNull();
+  }
+
+  @Test
+  public void closeStops() {
+    when(underlying.stopAsync()).thenReturn(underlying);
+    subscriber.close();
+    verify(underlying).stopAsync();
+    verify(underlying).awaitTerminated();
+  }
+
+  @Test
+  public void onDataAfterErrorThrows() {
+    CheckedApiException expected = new CheckedApiException(StatusCode.Code.INTERNAL);
+    errorListener.failed(null, expected);
+    ExecutionException e = assertThrows(ExecutionException.class, () -> subscriber.onData().get());
+    assertThat(expected).isEqualTo(e.getCause());
+  }
+
+  @Test
+  public void onDataBeforeErrorThrows() throws Exception {
+    CheckedApiException expected = new CheckedApiException(StatusCode.Code.INTERNAL);
+    Future<?> future =
+        executorService.submit(
```

Comment:
You don't need threading here. Just don't call get() until after calling .failed().